tensorboard/data/experimental/experiment_from_dev.py

```diff
@@ -0,0 +1,162 @@
+            "The get_scalars() feature requires the pandas package, "
+            "which does not seem to be available in your Python "
+            "environment. You can install it with command: \n\n"
+            "  pip install pandas"
+        )
+    return pandas
+
+
+class ExperimentFromDev(base_experiment.BaseExperiment):
+    """Implementation of BaseExperiment, specialized for tensorboard.dev."""
+
+    def __init__(self, experiment_id, api_endpoint=None):
+        super(ExperimentFromDev, self).__init__()
+        self._experiment_id = experiment_id
+        self._api_client = get_api_client(api_endpoint=api_endpoint)
+
+    def get_scalars(self, runs_filter=None, tags_filter=None, pivot=True):
+        if runs_filter is not None:
+            raise NotImplementedError(
+                "runs_filter support for get_scalars() is not implemented yet."
+            )
+        if tags_filter is not None:
+            raise NotImplementedError(
+                "tags_filter support for get_scalars() is not implemented yet."
+            )
+        pandas = import_pandas()
+
+        request = export_service_pb2.StreamExperimentDataRequest()
+        request.experiment_id = self._experiment_id
+        read_time = time.time()
+        util.set_timestamp(request.read_timestamp, read_time)
+        # TODO(cais, wchargin): Use another rpc to check for staleness and avoid
+        # a new StreamExperimentData rpc request if data is not stale.
+        stream = self._api_client.StreamExperimentData(
+            request, metadata=grpc_util.version_metadata()
+        )
+
+        runs = []
+        tags = []
+        steps = []
+        wall_times = []
+        values = []
+        for response in stream:
+            # TODO(cais, wchargin): Display progress bar.
+            metadata = base64.b64encode(
+                response.tag_metadata.SerializeToString()
+            ).decode("ascii")
+            num_values = len(response.points.values)
+            runs.extend([response.run_name] * num_values)
+            tags.extend([response.tag_name] * num_values)
+            steps.extend(list(response.points.steps))
+            wall_times.extend(
+                [t.ToNanoseconds() / 1e9 for t in response.points.wall_times]
+            )
+            values.extend(list(response.points.values))
+
+        dataframe = pandas.DataFrame(
+            {
+                "run": runs,
+                "tag": tags,
+                "step": steps,
+                "wall_time": wall_times,
+                "value": values,
+            }
+        )
+        if pivot:
+            dataframe = dataframe.pivot_table(
+                ["value", "wall_time"], ["run", "step"], "tag",
+            )
+            if np.any(dataframe.isnull().values):
+                sys.stderr.write(
+                    "WARNING: pivoted DataFrame contains missing value(s). "
+                    "This is likely due to tags that contain uneven numbers of steps. "
```

Comment:
This is a bit ambiguous, and I parsed it wrong the first time (and it
took me a while to figure out what was meant). I read “tags that contain
uneven numbers of steps” to mean “two tags with different numbers of
steps”, as in;

```python
# Tags with uneven numbers of steps: tag `acc` has 3 steps for run
# `train` but only 1 step for run `test`.
df = pandas.DataFrame(
    {
        "run": ["train", "train", "train", "test"],
        "tag": ["acc", "acc", "acc", "acc"],
        "step": [1, 2, 3, 3],
        "wall_time": [1000, 2000, 3000, 3500],
        "value": [0.1, 0.7, 0.9, 0.8],
    }
)
```

But this pivots without any problem. What’s actually meant is something
more like “one run for which the step values are not the same for all
tags in that run”. Like this:

```python
# Runs with uneven steps across tags: run `train` has steps 1, 2, and 3
# for tag `xent` but only step 3 for tag `xent`.
df = pandas.DataFrame(
    {
        "run": ["train", "train", "train", "train"],
        "tag": ["xent", "xent", "xent", "acc"],
        "step": [1, 2, 3, 3],
        "wall_time": [1000, 2000, 3000, 3500],
        "value": [0.1, 0.7, 0.9, 0.8],
    }
)
```

This example *does* have the problem, which is properly detected by the
code. Perhaps we can update the message to be clearer?